Add labels to release prs

josephperrott opened the pull request:
Automatically add provided labels to PRs created for release.

devversion commented on dev-infra/release/publish/actions.ts:
```diff
@@ -58,6 +59,9 @@ export interface ReleaseActionConstructor<T extends ReleaseAction = ReleaseActio
  */
 export abstract class ReleaseAction {
+  /** The target label to apply to the PR created for the action. */
+  abstract targetLabel: TargetLabel;
```
Would it make sense to don't tie the target label to an action? an action could send multiple PRs where different labels might be required (e.g. for the patch cherry pick)

I personally would vote for manually passing the label through. What do you think?

josephperrott replied:
Are you meaning having the user pass in the value? I think its important that the targeting is automatically determined the release tooling.

Passing through the label into the PR creation method seems much more reasonable to me.

However, an action that creates two PRs actually be two actions?

devversion replied:
I meant that we directly pass through the enum value when we call methods like `this.checkoutBranchAndStageVersion(newVersion`. A release action can create multiple PRs. e.g. one for changelog cherry picking.

That way we could also add an appropriate target label for such changelog cherry pick PRs.

josephperrott replied:
Ahh that makes sense. I can make that change.

josephperrott replied:
I ended up making a util function that gets the target label for a github branch.  And then just uses the targeted branch to determine the label.

devversion commented on dev-infra/release/publish/labels.ts:
```diff
@@ -0,0 +1,15 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+export enum TargetLabel {
```
Should this file live in the versioning folder instead? (as it's not specific to publishing). Also a comment refering to these labels as canonical org-wide ones might be good?

josephperrott replied:
Agreed, `versioning` is probably a better place for this.

gkalpak commented on dev-infra/release/publish/actions/cut-next-prerelease.ts:
```diff
@@ -20,6 +21,8 @@ import {ReleaseAction} from '../actions';
   /** Promise resolving with the new version if a NPM next pre-release is cut. */
   private _newVersion: Promise<semver.SemVer> = this._computeNewVersion();
+  /** The target label used for Pull Requests created during this action. */
+  targetLabel = TargetLabel.Major;
```
Why not `TargetLabel.Minor`? According to my understanding, `TargetLabel.Major` will be blocked (i.e. cannot be merged) unless the release train in the next branch is a new major version.

gkalpak commented on dev-infra/release/publish/actions/move-next-into-feature-freeze.ts:
```diff
@@ -22,6 +23,8 @@ import {packageJsonPath} from '../constants';
 export class MoveNextIntoFeatureFreezeAction extends ReleaseAction {
   private _newVersion = computeNewPrereleaseVersionForNext(this.active, this.config);
+  /** The target label used for Pull Requests created during this action. */
+  targetLabel = TargetLabel.Major;
```
According to my understanding, this should be `TargetLabel.RC`.
(If not `RC`, it should at least be `Minor`, but I believe it should be `RC` :grin:)

gkalpak commented on dev-infra/release/publish/actions.ts:
```diff
@@ -10,13 +10,16 @@ import {promises as fs} from 'fs';
 import {BuiltPackage, ReleaseConfig} from '../config';
 import {ActiveReleaseTrains} from '../versioning/active-release-trains';
 import {runNpmPublish} from '../versioning/npm-publish';
+import {TargetLabel} from '../versioning/target-labeling';
```
This doesn't seem to be used any more.

```suggestion
```

gkalpak commented on dev-infra/release/versioning/target-labeling.ts:
```diff
@@ -0,0 +1,20 @@
+  Patch = 'target: patch',
+  LTS = 'target: lts',
+  RC = 'target: rc',
+}
```
This file doesn't seem to be used any more. Should it be removed?

devversion commented on dev-infra/pr/merge/target-label.ts:
```diff
@@ -48,3 +48,20 @@ export async function getBranchesFromTargetLabel(
+          (typeof label.branches === 'function' ? await label.branches(branch) :
+                                                  await label.branches);
+      if (branches.length === 1 && branches[0] === branch) {
+        return label.pattern;
```
IIRC the pattern property could also be a RegExp. Maybe we should just hard-code the labels in each action. That might be simpler and a little more clear. I don't feel strongly about this though.

I'm just a bit unhappy that we'd evaluate the branches function unnecessarily, as it could potentially mean unecessary work (e.g. querying NPM/Github). Also the fact that we'd hide all types of runtime errors due to the try/catch. I personally liked the enum for the target labels a lot.

josephperrott replied:
I looked into this a bit and agree, but would like to do this in a follow up because it starts to get away from the scope of this change.  Additionally, it will likely bring on a bunch of import path changes because we can no longer rely on cross package implicit `/index` importing.

I will work on the follow up PR immediately though as I think its definitely important and worth doing right away.

gkalpak reviewed (changes requested):
The CI failures are legit ([example failure][1]), which I believe is related to one of the things @devversion mentioned in https://github.com/angular/angular/pull/40016#discussion_r545030420 (unnecessarily querying NPM/Github) :grin:

(Otherwise lgtm :+1:)

[1]: https://app.circleci.com/pipelines/github/angular/angular/26789/workflows/85259240-2f85-43c9-b98d-0e60b8f21779/jobs/892506
[2]: https://github.com/angular/angular/pull/40016#discussion_r545030420

josephperrott commented:
Ahh I see, I misunderstood what was being talked about with the github/npm checks.

I will step back and reassess how to go about everything.

josephperrott committed:
build: ensure action: merge label is included on release PRs

The release PRs should include the `action: merge`, `PullApprove: disable` and `comp: build & ci`
labels.

josephperrott committed:
feat(dev-infra): automatically add labels to release PRs

Automatically add provided labels to PRs created for release.

josephperrott commented:
@gkalpak Can you PTAL, I removed all of the target labeling portion and just left in the other label part for now.

gkalpak commented on .ng-dev/release.ts:
```diff
@@ -32,4 +32,7 @@ export const release: ReleaseConfig = {
   generateReleaseNotesForHead: async () => {
     exec('yarn -s gulp changelog', {cwd: join(__dirname, '../')});
   },
+  // The `comp: *` labels will be automatically added by PullApprove and the targeting label will
```
The comment about `comp: *` is not applicable any more.

gkalpak reviewed (approved):
:+1:

josephperrott committed:
fixup! feat(dev-infra): automatically add labels to release PRs
